### ISSUE TITLE
Allow custom logit warpers

### ISSUE THREAD

**Issue by lshamis** [huggingface/transformers#26191]
### Feature request

Currently, when generation sampling, logit warpers are automatically generated from the generation_config. 
https://github.com/huggingface/transformers/blob/0a55d9f7376f72ad3ff296d4249840021b03bcc4/src/transformers/generation/utils.py#L1641
which does not support any custom logit warpers
https://github.com/huggingface/transformers/blob/0a55d9f7376f72ad3ff296d4249840021b03bcc4/src/transformers/generation/utils.py#L811

Custom logits processor are allowed and follow a nearly identical code path.

### Motivation

I am unable to experiment with changes to sampling without also changing the logits via a LogitsProcessor.

### Your contribution

I can submit a PR, but am not necessarily the best person to do so.

**Comment by hedeershowk** [comment]
I see the same call in beam sampling and assisted generation.  Maybe custom logit warpers can be handled in `generate()` the same way `logits_processor` is used i.e., the caller can pass in a list of custom warpers as an arg `logits_warper` but has to be sure this doesn't collide with any from the config or raise?  I can take a shot at a PR if that seems like the right direction.

**Comment by ArthurZucker** [comment]
Hey @lshamis, you can probably overload `_get_logits_warper` in the mean time? The goal of transformers is to be easy to develop on top of it, but not necessarly fit your particular use case! If more people from the community require it, we might consider adding this. 
FYI @gante

**Comment by gante** [comment]
Hey @lshamis @hedeershowk 👋 

As @ArthurZucker wrote, we'd like to avoid adding more custom code if possible :) 

Let me attempt to present a solution that doesn't require new code: `logits_processor` and `logits_warper` behave the same way, the only difference being that the latter are only fetched on sampling-based decoding methods. However, nothing prevents you from adding a custom `logits_processor` for sampling purposes, e.g. the `top_k` operation. The only thing that you won't have control over is order of operations (if that's relevant for your use case), in which case I'd suggest to overwrite our functions as @ArthurZucker suggested!

I hope this helps 🤗
